File: c#-basics/patterns/iterator1.cs
namespace ConsoleApp18
{
    public class Book
    {
        public string Title { get; }

        public Book(string title)
        {
            Title = title;
        }
    }
    public class Library : IEnumerable<Book>
    {
        private List<Book> books = new List<Book>();

        public void AddBook(Book book)
        {
            books.Add(book);
        }

        public IEnumerator<Book> GetEnumerator()
        {
            foreach (var book in books)
            {
                yield return book;
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }
    }

    class Program
    {
        static void Main()
        {
            var library = new Library();
            library.AddBook(new Book("1984"));
            library.AddBook(new Book("Brave New World"));
            library.AddBook(new Book("Fahrenheit 451"));

            foreach (var book in library)
            {
                Console.WriteLine(book.Title);
            }
        }
    }
}

File: php-oop/patterns/iterator1.php
<?php

class BookCollection implements Iterator {
    private $books = [];
    private $index = 0;

    public function addBook(string $book) {
        $this->books[] = $book;
    }

    // Iterator methods
    public function current(): string {
        return $this->books[$this->index];
    }

    public function key(): int {
        return $this->index;
    }

    public function next(): void {
        $this->index++;
    }

    public function rewind(): void {
        $this->index = 0;
    }

    public function valid(): bool {
        return isset($this->books[$this->index]);
    }
}

// Użycie wzorca Iterator
$collection = new BookCollection();
$collection->addBook("1984");
$collection->addBook("Brave New World");
$collection->addBook("Fahrenheit 451");

foreach ($collection as $book) {
    echo $book . PHP_EOL;
}
